File: 19 09 25/spiralmatrix2.py
# LeetCode 59: Spiral Matrix II
# Explanation:
# 1. Start with an n x n matrix of zeros.
# 2. Maintain four boundaries: top, bottom, left, right.
# 3. Fill numbers in spiral order while shrinking boundaries inward.
# Time Complexity: O(n^2)
# Space Complexity: O(1) extra (besides output matrix)

class Solution:
    def generateMatrix(self, n: int):
        matrix = [[0]*n for _ in range(n)]
        left, right, top, bottom = 0, n-1, 0, n-1
        num = 1
        
        while left <= right and top <= bottom:
            for i in range(left, right+1):
                matrix[top][i] = num
                num += 1
            top += 1
            
            for i in range(top, bottom+1):
                matrix[i][right] = num
                num += 1
            right -= 1
            
            if top <= bottom:
                for i in range(right, left-1, -1):
                    matrix[bottom][i] = num
                    num += 1
                bottom -= 1
            
            if left <= right:
                for i in range(bottom, top-1, -1):
                    matrix[i][left] = num
                    num += 1
                left += 1
        
        return matrix
    

File: 19 09 25/spiralmatrix2.js
// LeetCode 59: Spiral Matrix II
// Explanation:
// 1. Track four boundaries (top,bottom,left,right) and fill numbers while spiraling inward.
// Time Complexity: O(n^2)

var generateMatrix = function(n) {
    const matrix = Array.from({length: n}, () => Array(n).fill(0));
    let top = 0, bottom = n-1, left = 0, right = n-1;
    let num = 1;
    
    while (left <= right && top <= bottom) {
        for (let i = left; i <= right; i++) matrix[top][i] = num++;
        top++;
        for (let i = top; i <= bottom; i++) matrix[i][right] = num++;
        right--;
        if (top <= bottom) {
            for (let i = right; i >= left; i--) matrix[bottom][i] = num++;
            bottom--;
        }
        if (left <= right) {
            for (let i = bottom; i >= top; i--) matrix[i][left] = num++;
            left++;
        }
    }
    return matrix;
};

